tensorboard/plugins/debugger_v2/debug_data_provider.py

```diff
@@ -95,11 +148,30 @@ def list_blob_sequences(
     def read_blob_sequences(
         self, experiment_id, plugin_name, downsample=None, run_tag_filter=None
     ):
-        del experiment_id, plugin_name, downsample, run_tag_filter  # Unused.
-        # TODO(cais): Implement this.
-        raise NotImplementedError()
+        del experiment_id, downsample  # Unused.
+        if plugin_name != PLUGIN_NAME:
+            raise ValueError("Unsupported plugin_name: %s" % plugin_name)
+        if not run_tag_filter.runs:
+            raise ValueError(
+                "run_tag_filter.runs is expected to be specified, but is not."
+            )
+
+        output = dict()
+        for run in run_tag_filter.runs:
+            if run in self._multiplexer.Runs():
+                output[run] = dict()
+            for tag in run_tag_filter.tags:
+                if tag.startswith(EXECUTION_DIGESTS_BLOB_TAG_PREFIX):
+                    output[run][tag] = [
+                        provider.BlobReference(blob_key="%s.%s" % (tag, run))
+                    ]
+        return output
 
     def read_blob(self, blob_key):
-        del blob_key  # Unused currently.
-        # TODO(cais): Implement this.
-        raise NotImplementedError()
+        if blob_key.startswith(EXECUTION_DIGESTS_BLOB_TAG_PREFIX):
+            run, begin, end = _parse_execution_digest_blob_key(blob_key)
```

Comment:
Just to make sure we’re on the same page: real data providers won’t be
able to dynamically generate blobs for particular plugins, so if you
want to be able to read from blobs identified by `(run, begin, end)`
then time series by these names will need to be created at upload time,
so `begin` and `end` will (probably) need to have fixed alignment and
fixed difference. For instance, perhaps `begin` is always a multiple of
10 and `end` is always `begin + 10`. We’re on the same page here, right?

Reply:
> In general, are implementations of DataProvider allowed to have
> internal logic beyond, say, requesting a whole object from GCS and
> passing it along to the caller without processing? Is the answer is
> yes, then begin and end can be arbitrary valid integer values, because
> the DataProvider impl and take care of making multiple requests to the
> backend and slicing/dicing the results to fulfill the given begin and
> end values.

They’re not expected to, no. The point of the data provider abstraction
is to decouple the storage layer from the plugin space, turning an
*m*-times-*n* problem into an *m*-plus-*n* problem. We don’t want to
write and maintain extra code for the debugger plugin for each storage
layer (local TensorBoard’s from-disk reading, TensorBoard.dev, any
internal TensorBoard hostd instances…). This is why I’ve been calling
out all the cases where the plugin is implicitly depending on undefined
or special-cased behavior from this data provider, because when we put
in a real data provider it _won’t_ have any debugger-specific code.

> The `DebuggerV2Plugin` just needs to do the work of making multiple
> `read_blob()` calls + slicing/dicing the results according to begin
> and end.

Yes, and as @stephanwlee pointed out, you’ll want to be careful about
the total number of `read_blob` RPCs needed to service one request here.